### ISSUE TITLE
[class.compare.default] p1 The condition of implicitly defining a defaulted comparison operator function CWG2546

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/draft#5336]
[class.compare.default] p1 just says
> A comparison operator function for class C that is defaulted on its first declaration and is not defined as deleted is [implicitly defined](https://eel.is/c++draft/class.compare#def:operator,comparison,implicitly_defined) when it is **odr-used** or **needed for constant evaluation**[.](https://eel.is/c++draft/class.compare#default-1.sentence-2)

Consider this example
````cpp
struct HasNoLessThan { };
struct C{
    friend HasNoLessThan operator <=>(C const&, C const&);
    bool operator <(C const&) const /*noexcept*/ = default;
};
int main(){
   using ptr = decltype(&C::operator<); // #1
}
````
Since `&C::operator<` appears as an unevaluated operand, function `C::operator <` is not odr-used by this expression. However, Clang reports a diagnosis that: 
> in evaluation of exception specification for 'C::operator<' needed here  

[except.spec] p11 says
> The exception specification for a comparison operator function ([over.binary]) without a noexcept-specifier that is defaulted on its first declaration is potentially-throwing if and only if **any expression in the implicit definition is potentially-throwing**.

So, It is reasonable to check the definition of the defaulted function. However, we just define two contexts that can cause the implicit definition for the defaulted function. Checking the expressions in the implicit definition is not defined to be odr-used nor constant evaluation of the function. It seems that we lack the case that determines the exception specification of a defaulted comparison operator function.

Improvement: 
> A comparison operator function for class C that is defaulted on its first declaration and is not defined as deleted is [implicitly defined](https://eel.is/c++draft/class.compare#def:operator,comparison,implicitly_defined) when it is odr-used, needed for constant evaluation, or needed to determine exception specification(if necessary)[.](https://eel.is/c++draft/class.compare#default-1.sentence-2)

**Comment by frederick-vs-ja** [comment]
[[dcl.fct.def.delete]/2](https://eel.is/c++draft/dcl.fct.def.delete#2)
> A program that refers to a deleted function implicitly or explicitly, other than to declare it, is ill-formed.

So even if the implicit exception specification is not needed, `decltype(&C::operator<)` is ill-formed.

IIIUC This is a bug of clang. Please report it [here](https://github.com/llvm/llvm-project/issues).

**Comment by xmh0511** [comment]
> [[dcl.fct.def.delete]/2](https://eel.is/c++draft/dcl.fct.def.delete#2)
> 
> > A program that refers to a deleted function implicitly or explicitly, other than to declare it, is ill-formed.
> 
> So even if the implicit exception specification is not needed, `decltype(&C::operator<)` is ill-formed.
> 
> IIUC this is a bug of clang. Please report it [here](https://github.com/llvm/llvm-project/issues).

No, the defaulted function is **not deleted**. `HasNoLessThan operator <=>(C const&, C const&)` is a usable candidate and a rewritten candidate as per [class.compare.secondary] p2

**Comment by frederick-vs-ja** [comment]
Are you explaining why the comment in [class.compare.secondary] p3 is wrong? I believe it's right, extremely.

When the rewritten candidate is found, a nested overload resolution is performed for the written expression and then fails. According to [[over.match.general] p4](https://eel.is/c++draft/over.match.general#4), the rewritten candidate is not usable.

**Comment by xmh0511** [comment]
> Are you explaining why the comment in [class.compare.secondary] p3 is wrong? I believe it's right, extremely.
> 
> When the rewritten candidate is found, a nested overload resolution is performed for the written expression and then fails. According to [[over.match.general] p4](https://eel.is/c++draft/over.match.general#4), the rewritten candidate is not usable.

"the rewritten candidate is not usable" No, it is usable. First, `HasNoLessThan operator <=>(C const&, C const&)` is a candidate for `x < y`, and all operands of the expression can be as the argument in the call of `HasNoLessThan operator <=>(C const&, C const&)` since the operands are of type `const C`. According to [over.match.general] p2 and [over.match.general] p3, `HasNoLessThan operator <=>(C const&, C const&)` is the unique best viable function. Hence, the overload resolution succeeds

**Comment by xmh0511** [comment]
If the comment is the intent of [class.compare.secondary] p2, it may be improved to
> The operator function with parameters x and y is defined as deleted if
>> - For expression x @ y, it is ill-formed, or
>> - the candidate selected by overload resolution, as applied to `x @ y`, is not a rewritten candidate.

**Comment by xmh0511** [comment]
> a nested overload resolution is performed for the written expression and then fails.

Another example that can prove we shouldn't augment the extent of considering overload resolution for this definition is:
````cpp
struct HasNoLessThan { 
    operator int(){
        return 0;
    }
};
struct C{
    friend HasNoLessThan operator <=>(C const&, C const&);
    bool operator <(C const&) const = default;
};
int main(){
   using ptr = decltype(&C::operator<);
}
````
Both implementation accept this example. The nested overload resolution will be performed for the operator `<` within `(x<=>y) < 0`. The selected candidate is the built-in one, which would violate the second bullet in [class.compare.secondary] p2. It is not reasonable to augment the extent of the overload resolution.

**Comment by jensmaurer** [comment]
The "selected candidate" in [class.compare.secondary] p2 is `oiperator <=>`, not the second-order overload resolution for `operator<`.

The fixes in CWG2546 should address this, too.

**Comment by frederick-vs-ja** [comment]
Should have been fixed by #6906 (eef4c2ba0d3952605e6ab839705f769690b92f7e).
